### PR TITLE
feat(list): add emptyState

### LINKED ISSUE #1653: [List] Add empty state
### What package is this for?

- [x] React
- [ ] Angular

### Summary

Add empty state to `List`.

**Additional context**
Currently, when not having any items inside the list, the list will simply shrink or be blank (if `isFullHeight` is set to `true`). Adding an empty state would help to signal the user that the list is currently empty and may be filled with content later on.

image

### PR DESCRIPTION
**Pull request by jonguenther**
Closes #1653 

**Summary**

- Add `emptyState` to be shown if `items` are empty and `hasEmptyState` is enabled
- `emptyState` can be either text or a component
- Add default `emptyState` style (see screenshot). If only text is provided, the text in the screenshot will be overwritten by it.

**Change List (commits, features, bugs, etc)**

- Add `emptyState` prop which will be shown if `List` is empty. Can be either text or component
- Add `hasEmptyState` prop which determines whether `emptyState` should be displayed if `items` is empty. `false` by default
- Add CSS for `emptyState` text styling
- Set defaultProp for `items` to `[]` and make it optional
- Add menu entry to `List` story

**Acceptance Test (how to verify the PR)**

- List shows no empty text if not set
- List shows empty text if desired
- Renders custom component for empty state

**Screenshot**
image

### CODE REVIEW

**Comment by jonguenther** [comment]
On a related note: Would it make sense to create a (internal) component for empty states? That way we would have a consistent look for it (or does such a component already exist?)

**Comment by tay1orjones** [comment]
@jonguenther good question - @dianatran18 have we thought about standardizing these empty states across the various components that use them?

**Comment by dianatran18** [comment]
Glad you asked @tay1orjones ! We actually just came out with some [empty state design guidance](https://pages.github.ibm.com/ai-applications/design/components/empty-states/usage/) to follow Carbon's standard more. 

Something I'd love to update in the PAL, and understand where these empty states are currently being used

File: src/components/List/List.jsx [hunk 1]
```diff
@@ -81,6 +82,10 @@ const propTypes = {
   onItemMoved: PropTypes.func,
   /** callback function when reorder will occur - can cancel the move by returning false */
   itemWillMove: PropTypes.func,
+  /** whether to show content if list is empty */
+  hasEmptyState: PropTypes.bool,
+  /** content shown if list is empty */
+  emptyState: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
```
Comment (tay1orjones) [inline src/components/List/List.jsx:88]: `hasEmptyState` seems redundant here. If `hasEmptyState` is removed,`emptyState` can be unconditionally rendered when there are no items.

If someone desires for the empty state to be totally blank, they can configure `emptyState` to be an empty string.

```suggestion
  /** content shown if list is empty */
  emptyState: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
```

File: src/components/List/List.jsx [hunk 2]
```diff
@@ -102,12 +107,15 @@ const defaultProps = {
...
   expandedIds: [],
+  items: [],
   handleSelect: () => {},
   toggleExpansion: () => {},
   onItemMoved: () => {},
   itemWillMove: () => {
     return true;
   },
+  hasEmptyState: false,
+  emptyState: 'Empty list',
```
Comment (tay1orjones) [inline src/components/List/List.jsx:118]:
```suggestion
  emptyState: 'No list items to show',
```

Loosely based on [the guidance from the site](https://pages.github.ibm.com/ai-applications/design/components/empty-states/usage#data-missing)

File: src/components/List/List.jsx [hunk 3]
```diff
@@ -236,7 +260,7 @@ const List = forwardRef((props, ref) => {
         )}
       >
         {!isLoading ? (
-          listItems
+          <>{listItems.length ? listItems : hasEmptyState && emptyContent}</>
```
Comment (tay1orjones) [inline src/components/List/List.jsx:263]:
```suggestion
          <>{listItems.length ? listItems : emptyContent}</>
```

File: src/components/List/List.story.jsx [hunk 4]
```diff
@@ -302,6 +302,18 @@ storiesOf('Watson IoT Experimental/List', module)
...
   ))
+  .add('with empty state', () => (
+    <div style={{ width: 400, height: 600 }}>
+      <List
+        title="Major League Baseball"
+        buttons={[headerButton]}
+        items={[]}
+        isLoading={boolean('isLoading', false)}
+        isFullHeight={boolean('isFullHeight', true)}
+        hasEmptyState={boolean('hasEmptyState', true)}
```
Comment (tay1orjones) [inline src/components/List/List.story.jsx:313]:
```suggestion
        emptyState={text('hasEmptyState', 'No list items to show')}
```

**Review by tay1orjones — CHANGES_REQUESTED** [review]
Knowing that we'll be looking to configure the new Empty state guidance Diana mentioned above, I think we can keep the API a bit simpler by having just one prop.

**Review by tay1orjones — APPROVED** [review]

**Review by davidicus — APPROVED** [review]
Good to go once CI checks have passed

**Comment by davidicus** [comment]
Looks like we just need to run yarn install for  prettier to pass the travis check

**Pull request merged**